cr-examples/spirv/src/main/java/intel/code/spirv/SpirvModuleGenerator.java

```diff
@@ -0,0 +1,1035 @@
+import java.lang.reflect.code.op.CoreOps;
+import java.lang.reflect.code.descriptor.MethodDesc;
+import java.lang.reflect.code.TypeElement;
+import java.lang.reflect.code.type.FunctionType;
+import java.lang.reflect.code.type.JavaType;
+import uk.ac.manchester.beehivespirvtoolkit.lib.InvalidSPIRVModuleException;
+import uk.ac.manchester.beehivespirvtoolkit.lib.SPIRVHeader;
+import uk.ac.manchester.beehivespirvtoolkit.lib.SPIRVModule;
+import uk.ac.manchester.beehivespirvtoolkit.lib.SPIRVFunction;
+import uk.ac.manchester.beehivespirvtoolkit.lib.SPIRVBlock;
+import uk.ac.manchester.beehivespirvtoolkit.lib.instructions.*;
+import uk.ac.manchester.beehivespirvtoolkit.lib.instructions.operands.*;
+import uk.ac.manchester.beehivespirvtoolkit.lib.disassembler.Disassembler;
+import uk.ac.manchester.beehivespirvtoolkit.lib.disassembler.SPIRVDisassemblerOptions;
+import uk.ac.manchester.beehivespirvtoolkit.lib.disassembler.CLIHighlighter;
+import uk.ac.manchester.beehivespirvtoolkit.lib.disassembler.SPIRVPrintingOptions;
+import uk.ac.manchester.beehivespirvtoolkit.lib.disassembler.SPVByteStreamReader;
+
+public class SpirvModuleGenerator {
+    public static MemorySegment generateModule(String moduleName, CoreOps.FuncOp func) {
+        SpirvOps.FuncOp spirvFunc = TranslateToSpirvModel.translateFunction(func);
+        MemorySegment module = SpirvModuleGenerator.generateModule(moduleName, spirvFunc);
+        return module;
+    }
+
+    public static MemorySegment generateModule(String moduleName, SpirvOps.FuncOp func) {
+        return new SpirvModuleGenerator().generateModuleInternal(moduleName, func);
+    }
+
+    public static void writeModuleToFile(MemorySegment module, String filepath)  {
+        ByteBuffer buffer = module.asByteBuffer();
+        File out = new File(filepath);
+        try (FileChannel channel = new FileOutputStream(out, false).getChannel()) {
+            channel.write(buffer);
+        }
+        catch (IOException e)  {
+            throw new RuntimeException(e);
+        }
+    }
+
+    public static String disassembleModule(MemorySegment module) {
+        SPVByteStreamReader input = new SPVByteStreamReader(new ByteArrayInputStream(module.toArray(ValueLayout.JAVA_BYTE)));
+        ByteArrayOutputStream out = new ByteArrayOutputStream();
+        try (PrintStream ps = new PrintStream(out))  {
+            SPIRVDisassemblerOptions options = new SPIRVDisassemblerOptions(false, false, false, false, true);
+            Disassembler dis = new Disassembler(input, ps, options);
+            dis.run();
+        }
+        catch (Exception e) {
+            throw new RuntimeException(e);
+        }
+        return new String(out.toByteArray());
+    }
+
+    private record SpirvResult(SPIRVId type, SPIRVId address, SPIRVId value) {}
+
+    private final SPIRVModule module;
+    private final Symbols symbols;
+
+    private SpirvModuleGenerator() {
+        this.module = new SPIRVModule(new SPIRVHeader(1, 2, 29, 0, 0));
```

Comment:
Suggestion: Since the SPIR-V is generated from the Beehive SPIR-V Toolkit, you can use the ID generator from that tool. The number is `32`, so it will be:

```java
this.module = new SPIRVModule(new SPIRVHeader(1, 2, 32, 0, 0));
```

Link: https://github.com/KhronosGroup/SPIRV-Headers/blob/main/include/spirv/spir-v.xml#L85